presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchangesForSingleNodeExecution.java

```diff
@@ -0,0 +1,150 @@
+import static com.facebook.presto.sql.planner.SystemPartitioningHandle.SINGLE_DISTRIBUTION;
+import static com.facebook.presto.sql.planner.iterative.rule.PickTableLayout.pushPredicateIntoTableScan;
+import static com.facebook.presto.sql.planner.plan.ExchangeNode.Scope.REMOTE_STREAMING;
+import static com.facebook.presto.sql.planner.plan.ExchangeNode.Type.GATHER;
+import static com.facebook.presto.sql.planner.plan.ExchangeNode.gatheringExchange;
+import static java.util.Objects.requireNonNull;
+
+public class AddExchangesForSingleNodeExecution
+        implements PlanOptimizer
+{
+    private final Metadata metadata;
+
+    public AddExchangesForSingleNodeExecution(Metadata metadata)
+    {
+        this.metadata = requireNonNull(metadata, "metadata is null");
+    }
+
+    @Override
+    public boolean isEnabled(Session session)
+    {
+        return isSingleNodeExecutionEnabled(session);
+    }
+
+    @Override
+    public PlanOptimizerResult optimize(PlanNode plan, Session session, TypeProvider types, VariableAllocator variableAllocator, PlanNodeIdAllocator idAllocator, WarningCollector warningCollector)
+    {
+        if (isEnabled(session)) {
+            AddExchangesForSingleNodeExecution.Rewriter rewriter = new AddExchangesForSingleNodeExecution.Rewriter(idAllocator, metadata, session);
+            PlanNode rewrittenPlan = SimplePlanRewriter.rewriteWith(rewriter, plan, null);
+            return PlanOptimizerResult.optimizerResult(rewrittenPlan, rewriter.isPlanChanged());
+        }
+        return PlanOptimizerResult.optimizerResult(plan, false);
+    }
+
+    private class Rewriter
+            extends SimplePlanRewriter<Void>
+    {
+        private final PlanNodeIdAllocator idAllocator;
+        private final Metadata metadata;
+        private final Session session;
+        private boolean planChanged;
+
+        private Rewriter(PlanNodeIdAllocator idAllocator, Metadata metadata, Session session)
+        {
+            this.idAllocator = requireNonNull(idAllocator, "idAllocator is null");
+            this.metadata = requireNonNull(metadata, "metadata is null");
+            this.session = requireNonNull(session, "session is null");
+        }
+
+        public boolean isPlanChanged()
+        {
+            return planChanged;
+        }
+
+        @Override
+        public PlanNode visitTableScan(TableScanNode node, RewriteContext<Void> context)
+        {
+            PlanNode plan = pushPredicateIntoTableScan(node, new ConstantExpression(true, BOOLEAN), true, session, idAllocator, metadata);
+            // Presto Java and Presto Native use different hash functions for partitioning
+            // An additional exchange makes sure the data flows through a native worker in case it need to be partitioned for downstream processing
+            if (containsSystemTableScan(plan)) {
+                plan = gatheringExchange(idAllocator.getNextId(), REMOTE_STREAMING, plan);
+            }
```

Comment:
Are these three lines of code the reason we've extended so many test cases to apply to Presto - single node - native?  If so, I'm wondering if a few, targetted example based tests are more appropriate.  I'm concerned the additional tests don't add value and will make our CI slower and more expensive.

Reply:
Not only so, essentially under single-node mode, we will not use addExchange optimizer. Theoretically we should test many cases that originally have remote exchange (especially partitioned), would they still return correct result under single-node mode.
 
And also scheduling has changes accordingly, one of the test cases actually caught a issue for it.

I understand your concern, I remove some of test (that I think could potentially be redundant in terms of exchange pattern, for example window function VS aggregation function)